clang/lib/Driver/ToolChains/DPURTE.cpp

```diff
@@ -82,51 +82,106 @@ void Linker::ConstructJob(Compilation &C, const JobAction &JA,
                           const llvm::opt::ArgList &TCArgs,
                           const char *LinkingOutput) const {
   std::string Linker = getToolChain().GetProgramPath(getShortName());
+
   // Put additional linker options
-  ArgStringList CmdArgs;
-  CmdArgs.push_back("--discard-locals");
+  ArgStringList LinkerCmdArgs;
+  AddLinkerInputs(getToolChain(), Inputs, TCArgs, LinkerCmdArgs, JA);
 
-  AddLinkerInputs(getToolChain(), Inputs, TCArgs, CmdArgs, JA);
-  CmdArgs.push_back("-o");
-  CmdArgs.push_back(Output.getFilename());
+  LinkerCmdArgs.push_back("--discard-locals");
+  LinkerCmdArgs.push_back("-o");
+  LinkerCmdArgs.push_back(Output.getFilename());
 
-  bool HasArgScript = false;
-  for (unsigned int EachArg = 0; EachArg < CmdArgs.size(); EachArg++) {
-    if (CmdArgs[EachArg][0] == '-' &&
-        (!strncmp("-T", CmdArgs[EachArg], 2) ||
-         !strncmp("--script", CmdArgs[EachArg], 8))) {
-      HasArgScript = true;
-      break;
-    }
-  }
-  if (!HasArgScript) {
-    CmdArgs.push_back("-T");
-    CmdArgs.push_back(LinkScript);
-  }
-
-  CmdArgs.push_back("-gc-sections");
+  LinkerCmdArgs.push_back("-gc-sections");
   // Must force common allocation, so that symbols with SHN_COMMON (aka .common)
   // have space allocated in WRAM. Otherwise, the linker places symbols at
   // the very beginning of memory with no allocation.
-  CmdArgs.push_back("--define-common");
+  LinkerCmdArgs.push_back("--define-common");
   if (!TCArgs.hasArg(options::OPT_nostdlib, options::OPT_nodefaultlibs)) {
     if (TCArgs.hasArg(options::OPT_flto) ||
         TCArgs.hasArg(options::OPT_flto_EQ)) {
       // Need to inject the RTE BC library into the whole chain.
-      CmdArgs.push_back(RtBcLibrary);
+      LinkerCmdArgs.push_back(RtBcLibrary);
     } else {
-      CmdArgs.push_back("-L");
-      CmdArgs.push_back(RtLibraryPath);
-      CmdArgs.push_back("-l");
-      CmdArgs.push_back(RtLibraryName);
+      LinkerCmdArgs.push_back("-L");
+      LinkerCmdArgs.push_back(RtLibraryPath);
+      LinkerCmdArgs.push_back("-l");
+      LinkerCmdArgs.push_back(RtLibraryName);
+    }
+  }
+
+  bool HasArgScript = false;
+  for (unsigned int EachArg = 0; EachArg < LinkerCmdArgs.size(); EachArg++) {
+    if ((LinkerCmdArgs[EachArg][0] == '-') &&
+        (!strncmp("-T", LinkerCmdArgs[EachArg], 2) ||
+         !strncmp("--script", LinkerCmdArgs[EachArg], 8))) {
+      HasArgScript = true;
+      // break;
+    }
+  }
+
+  if (!HasArgScript) {
+    LinkerCmdArgs.push_back("-T");
+    LinkerCmdArgs.push_back(LinkScript);
+  }
+
+#define STR_BUFFER_SIZE 128
+#define NR_THREADS 24
```

Comment:
We should probably have a generated file somewhere with the DPU characteristics (maybe not for this PR, but in general).

Reply:
agreed